src/ds/openapi.h

```diff
@@ -0,0 +1,393 @@
+            "registered with this name: {} vs {}",
+            name,
+            schema_.dump(),
+            existing_schema.dump()));
+        }
+      }
+      else
+      {
+        schemas.emplace(name, schema_);
+      }
+
+      return components_ref_object(name);
+    }
+
+    struct SchemaHelper
+    {
+      nlohmann::json& document;
+
+      template <typename T>
+      nlohmann::json add_schema_component()
+      {
+        nlohmann::json schema;
+        if constexpr (nonstd::is_specialization<T, std::optional>::value)
+        {
+          return add_schema_component<typename T::value_type>();
+        }
+        else if constexpr (nonstd::is_specialization<T, std::vector>::value)
+        {
+          schema["type"] = "array";
+          schema["items"] = add_schema_component<typename T::value_type>();
+
+          return add_schema_to_components(
+            document, ds::json::schema_name<T>(), schema);
+        }
+        else if constexpr (
+          nonstd::is_specialization<T, std::map>::value ||
+          nonstd::is_specialization<T, std::unordered_map>::value)
+        {
+          // Nlohmann serialises maps to an array of (K, V) pairs
+          if (std::is_same<typename T::key_type, std::string>::value)
+          {
+            // ...unless the keys are strings!
+            schema["type"] = "object";
+            schema["additionalProperties"] =
+              add_schema_component<typename T::mapped_type>();
+          }
+          else
+          {
+            schema["type"] = "array";
+            auto items = nlohmann::json::object();
+            {
+              items["type"] = "array";
+
+              auto sub_items = nlohmann::json::array();
+              // NB: OpenAPI doesn't like this tuple for "items", even though
+              // its valid JSON schema. May need to switch this to oneOf to
+              // satisfy some validators
+              sub_items.push_back(add_schema_component<typename T::key_type>());
+              sub_items.push_back(
+                add_schema_component<typename T::mapped_type>());
+              items["items"] = sub_items;
+            }
+            schema["items"] = items;
+          }
+          return add_schema_to_components(
+            document, ds::json::schema_name<T>(), schema);
+        }
+        else if constexpr (nonstd::is_specialization<T, std::pair>::value)
+        {
+          schema["type"] = "array";
+          auto items = nlohmann::json::array();
+          items.push_back(add_schema_component<typename T::first_type>());
+          items.push_back(add_schema_component<typename T::second_type>());
+          schema["items"] = items;
+          return add_schema_to_components(
+            document, ds::json::schema_name<T>(), schema);
+        }
+        else if constexpr (
+          std::is_same<T, std::string>::value || std::is_same<T, bool>::value ||
+          std::is_same<T, uint8_t>::value || std::is_same<T, uint16_t>::value ||
```

Comment:
This seems like it could be shorter if it used is_arithmetic (unless I'm missing something we want to exclude explicitly?)

Reply:
Good suggestion! This was a copy-paste reduction from `json_schema.h`, where we want to name each type individually, but `is_arithmetic` is much shorter.